gecoscc/utils.py

```diff
@@ -2245,3 +2261,259 @@ def trace_inheritance(logger, db, action, obj, policy):
+    ''' Backing up all Chef server data
+    
+    Args:
+      username(str):    user with permission in chef server
+      backupdir(str):   backup directory where hold files
+    '''
+    logger.info("Backing up Chef Server ...")
+
+    try:
+ 
+        settings = get_current_registry().settings
+
+        if is_cli_request():
+            has_cli_permission(os.environ['SCRIPT_CODE'], chefserver_backup.__name__)
+            username = pickle.loads(os.environ['GECOS_USER']).get('username',None)
+            backupdir = os.environ['BACKUP_DIR']
+
+        logger.debug("utils.py ::: chefserver_backup - username = %s" % username)
+        logger.debug("utils.py ::: chefserver_backup - backupdir = %s" % backupdir)
+
+        assert username is not None or backupdir is not None, _('Missing required arguments')
+
+        if not os.path.exists(backupdir):
+            os.mkdir(backupdir)
+
+        admin_cert = os.sep.join([settings.get('firstboot_api.media'), username, 'chef_user.pem'])
+        logger.debug("utils.py ::: chefserver_backup - admin_cert = %s" % admin_cert)
+
+        command = 'source /opt/rh/rh-ruby24/enable; knife backup export -D {0} -y -s {1} -u {2} -k {3}'.format(backupdir, settings.get('chef.url'), username, admin_cert)
+        backup_output = subprocess.check_output(command, shell=True)
+        logger.info(backup_output)
+        logger.info("Chef Server backup ended.")
+
+    except AssertionError, msg:
+        logger.error(msg)
+
+    except subprocess.CalledProcessError, msg:
+        logger.error(msg.cmd)
+        logger.error(msg.output)
+
+def chefserver_restore(username=None, backupdir=None):
+    ''' Restoring Chef server data from a backup that was created by the chefserver_backup function
+    
+    Args:
+      username(str):    user with permission in chef server
+      backupdir(str):   directory where backup files are
+    '''
+    logger.info("Restoring Chef Server ...")
+
+    try:
+
+        settings = get_current_registry().settings
+
+        if is_cli_request():
+            has_cli_permission(os.environ['SCRIPT_CODE'], chefserver_restore.__name__)
+            username = pickle.loads(os.environ['GECOS_USER']).get('username',None)
+            backupdir = os.environ['BACKUP_DIR']
+
+        logger.debug("utils.py ::: chefserver_backup - username = %s" % username)
+        logger.debug("utils.py ::: chefserver_backup - backupdir = %s" % backupdir)
+
+        assert username is not None or backupdir is not None, _('Missing required arguments')
+
+        if not os.path.exists(backupdir):
+            os.mkdir(backupdir)
+
+        admin_cert = os.sep.join([settings.get('firstboot_api.media'), username, 'chef_user.pem'])
+        logger.debug("utils.py ::: chefserver_backup - admin_cert = %s" % admin_cert)
+
+        command = 'source /opt/rh/rh-ruby24/enable; knife backup restore -D {0} -y -s {1} -u {2} -k {3}'.format(backupdir, settings.get('chef.url'), username, admin_cert)
```

Comment:
Please, see last comment

Reply:
It's done. I've added a function called `source`, that emulates BASH command `source` to set environment variables

Reply:
The previous solution is discarded and replaced by the execution of script that configures the ruby environment and executes the knife command